# 1 Assignment 3

Diego Delgado Díaz
Maria del Carmen Venero
Gonzalo Canelo
Luis Quispe

In [30]:
import os
import os
import pickle
import pandas as pd
import numpy as np
import urllib.request
import savReaderWriter as sav

1. Import the `REC0111.sav`, `RE223132.sav` and `RE516171.sav` files and their variables and values labels from this path `"../../_data/endes/2019"`. The name of imported files should be named as `rec_1`, `rec_2` and `rec_3` for files `REC0111.sav`, `RE223132.sav` and `RE516171.sav` respectively. The name of the variable and value labels should be `var_labels1` and `value_labels1` for `rec1`, `var_labels2` and `value_labels2` for `rec2`, and `var_labels3` and `value_labels3` for `rec3`. **Hint: See the section 3.3.4 of [the lecture 3](https://github.com/alexanderquispe/Diplomado_PUCP/blob/main/Lecture_3/Lecture_3.ipynb)**

## Solution:

We use the pd.read_spss() to import the data. We generate 3 dataframes for each dataset.

In [31]:
rec_1 = pd.read_spss( r"../../_data/endes/2019/REC0111.sav" )
rec_2 = pd.read_spss( r"../../_data/endes/2019/RE223132.sav" )
rec_3 = pd.read_spss( r"../../_data/endes/2019/RE516171.sav" )

We check at the data

In [32]:
rec_1

ID1             HHID              CASEID    V001  V002  V003  \
0      2019.0        000100201        000100201  2     1.0   2.0   2.0   
1      2019.0        000100201        000100201  3     1.0   2.0   3.0   
2      2019.0        000102801        000102801  2     1.0  28.0   2.0   
3      2019.0        000102801        000102801  6     1.0  28.0   6.0   
4      2019.0        000104801        000104801  2     1.0  48.0   2.0   
...       ...              ...                 ...     ...   ...   ...   
38330  2019.0        325406201        325406201  2  3254.0  62.0   2.0   
38331  2019.0        325406301        325406301  2  3254.0  63.0   2.0   
38332  2019.0        325407001        325407001  2  3254.0  70.0   2.0   
38333  2019.0        325407201        325407201  2  3254.0  72.0   2.0   
38334  2019.0        325407401        325407401  2  3254.0  74.0   2.0   

         V004    V007    V008  V009  ...  QD333_4  QD333_5  QD333_6  UBIGEO  \
0         1.0  2019.0  1434.0   4.0  ...       No       No       No  010101   
1         1.0  2019.0  1434.0   1.0  ...       No       No       No  010101   
2         1.0  2019.0  1434.0   6.0  ...       No       No       No  010101   
3         1.0  2019.0  1434.0   3.0  ...       No       No       No  010101   
4         1.0  2019.0  1434.0   5.0  ...       No       No       No  010101   
...       ...     ...     ...   ...  ...      ...      ...      ...     ...   
38330  3254.0  2019.0  1440.0  12.0  ...       No       No       No  250401   
38331  3254.0  2019.0  1440.0   6.0  ...       No       No       No  250401   
38332  3254.0  2019.0  1440.0   7.0  ...       No       No       No  250401   
38333  3254.0  2019.0  1440.0  12.0  ...       No       No       No  250401   
38334  3254.0  2019.0  1440.0  10.0  ...       No       No       No  250401   

        V022      V005          V190      V191  \
0        3.0  154803.0          Rico  1.234450   
1        3.0  154803.0          Rico  1.234450   
2        3.0  154803.0          Rico  1.295611   
3        3.0  154803.0          Rico  1.295611   
4        3.0  154803.0        Pobrer -0.256431   
...      ...       ...           ...       ...   
38330  249.0  244995.0  El más pobre -1.750187   
38331  249.0  244995.0  El más pobre -1.676861   
38332  249.0  459792.0  El más pobre -1.585333   
38333  249.0  459792.0  El más pobre -1.650159   
38334  249.0  244995.0  El más pobre -1.644720   

                                        mujeres12a49  NCONGLOME  
0                    Mujeres de 15 a 49 años de edad     7065.0  
1      Mujeres de 12 a 14 de edad, nunca embarazadas     7065.0  
2                    Mujeres de 15 a 49 años de edad     7065.0  
3                    Mujeres de 15 a 49 años de edad     7065.0  
4                    Mujeres de 15 a 49 años de edad     7065.0  
...                                              ...        ...  
38330                Mujeres de 15 a 49 años de edad    15783.0  
38331                Mujeres de 15 a 49 años de edad    15783.0  
38332                Mujeres de 15 a 49 años de edad    15783.0  
38333                Mujeres de 15 a 49 años de edad    15783.0  
38334                Mujeres de 15 a 49 años de edad    15783.0  

[38335 rows x 105 columns]

In [33]:
rec_2

ID1              CASEID  V201  V202  V203  V204  V205  V206  V207  \
0      2019.0        000100201  2   2.0   1.0   1.0   0.0   0.0   0.0   0.0   
1      2019.0        000100201  3   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
2      2019.0        000102801  2   3.0   1.0   2.0   0.0   0.0   0.0   0.0   
3      2019.0        000102801  6   3.0   1.0   0.0   2.0   0.0   0.0   0.0   
4      2019.0        000104801  2   3.0   1.0   2.0   0.0   0.0   0.0   0.0   
...       ...                 ...   ...   ...   ...   ...   ...   ...   ...   
36917  2019.0        325406201  2   5.0   1.0   1.0   2.0   1.0   0.0   0.0   
36918  2019.0        325406301  2   5.0   2.0   2.0   0.0   1.0   0.0   0.0   
36919  2019.0        325407001  2   8.0   2.0   0.0   5.0   0.0   0.0   1.0   
36920  2019.0        325407201  2   2.0   1.0   0.0   0.0   0.0   1.0   0.0   
36921  2019.0        325407401  2   1.0   1.0   0.0   0.0   0.0   0.0   0.0   

       V208  ...  V307_09  V307_10  V307_11  V307_12 V307_13  V307_14 V307_15  \
0       1.0  ...      NaN      NaN      NaN      NaN     NaN      NaN     NaN   
1       0.0  ...      NaN      NaN      NaN      NaN     NaN      NaN     NaN   
2       2.0  ...       No       No      NaN      NaN     NaN      NaN     NaN   
3       0.0  ...      NaN      NaN      NaN      NaN     NaN      NaN     NaN   
4       1.0  ...      NaN      NaN      NaN      NaN     NaN      NaN     NaN   
...     ...  ...      ...      ...      ...      ...     ...      ...     ...   
36917   1.0  ...      NaN      NaN      NaN      NaN     NaN      NaN     NaN   
36918   1.0  ...       No       Sí      NaN      NaN     NaN      NaN     NaN   
36919   0.0  ...       Sí       No      NaN      NaN     NaN      NaN     NaN   
36920   0.0  ...      NaN      NaN      NaN      NaN     NaN      NaN     NaN   
36921   1.0  ...       No      NaN      NaN      NaN     NaN      NaN     NaN   

      V307_16 QI302_05A  QI302_05B  
0          No        Sí         No  
1         NaN       NaN        NaN  
2          No        No         Sí  
3         NaN        No         No  
4         NaN        No         Sí  
...       ...       ...        ...  
36917     NaN        No         Sí  
36918     NaN        No         Sí  
36919     NaN        No         Sí  
36920     NaN        No         Sí  
36921     NaN        No         Sí  

[36922 rows x 176 columns]

In [34]:
rec_3

ID1              CASEID             V501                  V502  \
0      2019.0        000100201  2           Casado    Actualmente casada   
1      2019.0        000102801  2  Viviendo juntos    Actualmente casada   
2      2019.0        000102801  6  No viven juntos  Anteriormente casada   
3      2019.0        000104801  2  Viviendo juntos    Actualmente casada   
4      2019.0        000113601  2  Viviendo juntos    Actualmente casada   
...       ...                 ...              ...                   ...   
33306  2019.0        325406201  2  Viviendo juntos    Actualmente casada   
33307  2019.0        325406301  2  Viviendo juntos    Actualmente casada   
33308  2019.0        325407001  2  Viviendo juntos    Actualmente casada   
33309  2019.0        325407201  2  Viviendo juntos    Actualmente casada   
33310  2019.0        325407401  2  Viviendo juntos    Actualmente casada   

                 V503           V504 V505 V506  V507    V508  ...  \
0             Una vez  Vive con ella  NaN  NaN   1.0  2008.0  ...   
1             Una vez  Vive con ella  NaN  NaN  12.0  2011.0  ...   
2             Una vez            NaN  NaN  NaN   8.0  1984.0  ...   
3             Una vez  Vive con ella  NaN  NaN  12.0  2009.0  ...   
4      Mas de una vez  Vive con ella  NaN  NaN  12.0  2005.0  ...   
...               ...            ...  ...  ...   ...     ...  ...   
33306  Mas de una vez  Vive con ella  NaN  NaN  12.0  1985.0  ...   
33307         Una vez  Vive con ella  NaN  NaN  12.0  2002.0  ...   
33308         Una vez  Vive con ella  NaN  NaN  11.0  1986.0  ...   
33309         Una vez  Vive con ella  NaN  NaN   5.0  2008.0  ...   
33310         Una vez  Vive con ella  NaN  NaN   6.0  2013.0  ...   

                                 V743C                            V743D  \
0                         Entrevistada  Entrevistada y esposo/compañero   
1      Entrevistada y esposo/compañero                     Entrevistada   
2                                  NaN                              NaN   
3                         Entrevistada                     Entrevistada   
4                         Entrevistada                     Entrevistada   
...                                ...                              ...   
33306  Entrevistada y esposo/compañero  Entrevistada y esposo/compañero   
33307                     Entrevistada                     Entrevistada   
33308                 Esposo/compañero                 Esposo/compañero   
33309                 Esposo/compañero                     Entrevistada   
33310                 Esposo/compañero                 Esposo/compañero   

                                 V743E                            V743F V744A  \
0                         Entrevistada                 Esposo/compañero    No   
1                         Entrevistada  Entrevistada y esposo/compañero    No   
2                                  NaN                              NaN    No   
3                         Entrevistada                     Entrevistada    No   
4      Entrevistada y esposo/compañero                 Esposo/compañero    No   
...                                ...                              ...   ...   
33306  Entrevistada y esposo/compañero  Entrevistada y esposo/compañero    No   
33307                     Entrevistada  Entrevistada y esposo/compañero    No   
33308                     Entrevistada  Entrevistada y esposo/compañero    No   
33309                     Entrevistada  Entrevistada y esposo/compañero    No   
33310                     Entrevistada  Entrevistada y esposo/compañero    No   

      V744B V744C V744D V744E          V746  
0        No    No    No    No    Más que el  
1        No    No    No    No  Menos que el  
2        No    No    No    No           NaN  
3        No    No    No    No           NaN  
4        No    No    No    No  Menos que el  
...     ...   ...   ...   ...           ...  
33306    No    No    No    No           NaN  
33307    No    No    No   

Then we import the labels for each dataset.

In [35]:
# for rec_1
with sav.SavHeaderReader( r"../../_data/endes/2019/REC0111.sav", ioUtf8=True) as header:
    metadata1 = header.all()
    labels_rec1 = metadata1.valueLabels
    
# for rec_2
with sav.SavHeaderReader( r"../../_data/endes/2019/RE223132.sav", ioUtf8=True) as header:
    metadata2 = header.all()
    labels_rec2 = metadata2.valueLabels
    
# for rec_3
with sav.SavHeaderReader( r"../../_data/endes/2019/RE516171.sav", ioUtf8=True) as header:
    metadata3 = header.all()
    labels_rec3 = metadata3.valueLabels    

We check at the dictionaries

In [36]:
labels_rec1

{'V013': {0.0: 'De 12 a 14 años de edad',
  1.0: 'De 15 a 19 años de edad',
  2.0: 'De 20 a 24 años de edad',
  3.0: 'De 25 a 29 años de edad',
  4.0: 'De 30 a 34 años de edad',
  5.0: 'De 35 a 39 años de edad',
  6.0: 'De 40 a 44 años de edad',
  7.0: 'De 45 a 49 años de edad'},
 'V014': {1.0: 'Mes y año',
  2.0: 'Mes y edad - año imputado',
  3.0: 'Año y edad - mes imputado',
  4.0: 'Año y edad - año ignorado supera el año de entrevista',
  5.0: 'Año - edad y mes imputados',
  6.0: 'Edad - año y mes imputados',
  7.0: 'Mes - edad y año imputados',
  8.0: 'Ninguno - edad, mes y año imputados'},
 'V015': {1.0: 'Completa',
  2.0: 'Ausente',
  3.0: 'Aplazada',
  4.0: 'Rechazada',
  5.0: 'Incoompleta',
  6.0: 'Discapacitada',
  7.0: 'Otra'},
 'V018': {0.0: 'Sin calendario'},
 'V019': {0.0: 'Sin calendario'},
 'V020': {0.0: 'Toda la mujer muestra', 1.0: 'Alguna muestra casada'},
 'V023': {1.0: 'Amazonas',
  2.0: 'Ancash',
  3.0: 'Apurimac',
  4.0: 'Arequipa',
  5.0: 'Ayacucho',
  6.0: 'Caj

In [37]:
labels_rec2

{'V213': {0.0: 'No o no está segura', 1.0: 'Sí'},
 'V215': {994.0: 'En la menopausia',
  995.0: 'Antes del último nacimiento',
  996.0: 'Nunca menstruado',
  997.0: 'Inconsistente',
  998.0: 'No sabe'},
 'V216': {0.0: 'No', 1.0: 'Si'},
 'V217': {1.0: 'Durante su periodo',
  2.0: 'Después de su período',
  3.0: 'En medio del ciclo',
  4.0: 'Antes del período',
  5.0: 'En cualquier momento',
  6.0: 'Otro',
  8.0: 'No sabe'},
 'V220': {6.0: '6+'},
 'V221': {996.0: 'Intervalo Negativo'},
 'V223': {7.0: 'Mes - fecha exacta', 8.0: 'Ninguno - fecha de colocación'},
 'V225': {1.0: 'En ese momento',
  2.0: 'Querìa esperar',
  3.0: 'No querìa tener (màs) hijas/hijos'},
 'V226': {994.0: 'En la menopausia',
  995.0: 'Antes del último nacimiento',
  996.0: 'Nunca menstruado',
  997.0: 'Inconsistente',
  998.0: 'No sabe'},
 'V227': {0.0: 'No marcado',
  1.0: 'N> int. desde el nacimiento',
  2.0: 'N + amen > intervalo',
  3.0: 'N, pero per no ret',
  4.0: '95, pero sin nacimiento',
  5.0: '95, pero p

In [38]:
labels_rec3

{'V501': {0.0: 'Nunca casada',
  1.0: 'Casado',
  2.0: 'Viviendo juntos',
  3.0: 'Viuda',
  4.0: 'Divorciada',
  5.0: 'No viven juntos'},
 'V502': {0.0: 'Nunca casada',
  1.0: 'Actualmente casada',
  2.0: 'Anteriormente casada'},
 'V503': {1.0: 'Una vez', 2.0: 'Mas de una vez'},
 'V504': {1.0: 'Vive con ella', 2.0: 'Vive en otro sitio'},
 'V505': {0.0: 'No otras esposas', 98.0: 'No sabe'},
 'V506': {98.0: 'No sabe'},
 'V510': {1.0: 'Tiene mes y año',
  2.0: 'Tiene mes y edad - imputa año',
  3.0: 'Tiene año y edad - imputa mes',
  5.0: 'Tiene año - imputa mes',
  6.0: 'Tiene edad - imputa mes año',
  7.0: 'Tiene mes - imputa año',
  8.0: 'Ninguno - imputa todo'},
 'V513': {0.0: 'Nunca casada',
  1.0: 'De 0 a 4 años',
  2.0: 'De 5 a 9 años',
  3.0: 'De 10 a 14 años',
  4.0: 'De 15 a 19 años',
  5.0: 'De 20 a 24 años',
  6.0: 'De 25 a 29 años',
  7.0: 'De 30 a más años'},
 'V525': {0.0: 'Nunca tuvo relaciones sexuales', 96.0: 'En la primera union'},
 'V527': {100.0: 'Menos de un día',
  

Now we add the labels to variables

In [39]:
# Labels for variables in rec_1 
rec_1.attrs[ 'value_labels1' ] = labels_rec1
rec_1.attrs[ 'var_labels1' ] = metadata1.varLabels

# Labels for variables in rec_2 
rec_2.attrs[ 'value_labels2' ] = labels_rec2
rec_2.attrs[ 'var_labels2' ] = metadata2.varLabels

# Labels for variables in rec_3 
rec_3.attrs[ 'value_labels3' ] = labels_rec3
rec_3.attrs[ 'var_labels3' ] = metadata3.varLabels

In [40]:
rec_1.attrs[ 'value_labels1' ]
rec_2.attrs[ 'value_labels2' ]
rec_3.attrs[ 'value_labels3' ]

{'V501': {0.0: 'Nunca casada',
  1.0: 'Casado',
  2.0: 'Viviendo juntos',
  3.0: 'Viuda',
  4.0: 'Divorciada',
  5.0: 'No viven juntos'},
 'V502': {0.0: 'Nunca casada',
  1.0: 'Actualmente casada',
  2.0: 'Anteriormente casada'},
 'V503': {1.0: 'Una vez', 2.0: 'Mas de una vez'},
 'V504': {1.0: 'Vive con ella', 2.0: 'Vive en otro sitio'},
 'V505': {0.0: 'No otras esposas', 98.0: 'No sabe'},
 'V506': {98.0: 'No sabe'},
 'V510': {1.0: 'Tiene mes y año',
  2.0: 'Tiene mes y edad - imputa año',
  3.0: 'Tiene año y edad - imputa mes',
  5.0: 'Tiene año - imputa mes',
  6.0: 'Tiene edad - imputa mes año',
  7.0: 'Tiene mes - imputa año',
  8.0: 'Ninguno - imputa todo'},
 'V513': {0.0: 'Nunca casada',
  1.0: 'De 0 a 4 años',
  2.0: 'De 5 a 9 años',
  3.0: 'De 10 a 14 años',
  4.0: 'De 15 a 19 años',
  5.0: 'De 20 a 24 años',
  6.0: 'De 25 a 29 años',
  7.0: 'De 30 a más años'},
 'V525': {0.0: 'Nunca tuvo relaciones sexuales', 96.0: 'En la primera union'},
 'V527': {100.0: 'Menos de un día',
  

2. Select the following columns for each data set:
|Data|Columns|
|---|---|
|rec1| CASEID, V000, V001, V002, V003, V004, V007, V008, V009, V010, V011, V012, V024, V102, V120, V121, V122, V123, V124, V125, V127, V133 |
|rec2| CASEID, V201, V218, V301, V302, V323, V323A, V325A, V326, V327, V337, V359, V360, V361, V362, V363, V364, V367, V372, V372A, V375A, V376, V376A, V379, V380 |
|rec3| CASEID, V501, V502, V503, V504, V505, V506, V507, V508, V509, V510, V511, V512, V513, V525, V613, V714, V715 |


Additioanlly, you should update the variables and value labels objects. They must have information only for the selected columns. The new dataframes should be name as `rec1_1`, `rec2_1`, and `rec3_1`. The new varible labels objects should be named as `new_var_labels1`, `new_var_labels2`, and `new_var_labels3`. The new value labels objects should be named as `new_value_labels1`, `new_value_labels2`, and `new_value_labels3` **Hint: Use the `loc` and column names to filter. Update the dictionary.**

In [41]:
rec1_1 = rec_1.loc[:, ['CASEID', 'V000', 'V001', 'V002', 'V003', 'V004', 'V007', 'V008', 'V009', 'V010', 'V011', 'V012', 'V024', 'V102', 'V120', 'V121', 'V122', 'V123', 'V124', 'V125', 'V127', 'V133']]
rec1_1

CASEID V000    V001  V002  V003    V004    V007    V008  \
0            000100201  2  PE6     1.0   2.0   2.0     1.0  2019.0  1434.0   
1            000100201  3  PE6     1.0   2.0   3.0     1.0  2019.0  1434.0   
2            000102801  2  PE6     1.0  28.0   2.0     1.0  2019.0  1434.0   
3            000102801  6  PE6     1.0  28.0   6.0     1.0  2019.0  1434.0   
4            000104801  2  PE6     1.0  48.0   2.0     1.0  2019.0  1434.0   
...                   ...  ...     ...   ...   ...     ...     ...     ...   
38330        325406201  2  PE6  3254.0  62.0   2.0  3254.0  2019.0  1440.0   
38331        325406301  2  PE6  3254.0  63.0   2.0  3254.0  2019.0  1440.0   
38332        325407001  2  PE6  3254.0  70.0   2.0  3254.0  2019.0  1440.0   
38333        325407201  2  PE6  3254.0  72.0   2.0  3254.0  2019.0  1440.0   
38334        325407401  2  PE6  3254.0  74.0   2.0  3254.0  2019.0  1440.0   

       V009    V010  ...      V024    V102 V120 V121 V122 V123 V124 V125  \
0       4.0  1986.0  ...  Amazonas  Urbano   Si   Si   Si   No   No   No   
1       1.0  2007.0  ...  Amazonas  Urbano   Si   Si   Si   No   No   No   
2       6.0  1983.0  ...  Amazonas  Urbano   Si   Si   Si   Si   Si   Si   
3       3.0  1970.0  ...  Amazonas  Urbano   Si   Si   Si   Si   Si   Si   
4       5.0  1991.0  ...  Amazonas  Urbano   No   No   No   No   No   No   
...     ...     ...  ...       ...     ...  ...  ...  ...  ...  ...  ...   
38330  12.0  1971.0  ...   Ucayali   Rural   No   No   No   No   No   No   
38331   6.0  1988.0  ...   Ucayali   Rural   No   No   No   No   No   No   
38332   7.0  1973.0  ...   Ucayali   Rural   No   No   No   No   No   No   
38333  12.0  1994.0  ...   Ucayali   Rural   No   No   No   No   No   No   
38334  10.0  1996.0  ...   Ucayali   Rural   No   No   No   No   No   No   

                                V127  V133  
0      Losetas, terrazos o similares  16.0  
1      Losetas, terrazos o similares   6.0  
2      Losetas, terrazos o similares  16.0  
3      Losetas, terrazos o similares   4.0  
4                   Cemento/ladrillo   1.0  
...                              ...   ...  
38330                           Pona   4.0  
38331                           Pona   6.0  
38332                           Pona   3.0  
38333                           Pona   9.0  
38334                           Pona   9.0  

[38335 rows x 22 columns]

In [42]:
rec2_1 = rec_2.loc[:, ['CASEID', 'V201', 'V218', 'V301', 'V302', 'V323', 'V323A', 'V325A', 'V326', 'V327', 'V337', 'V359', 'V360', 'V361', 'V362', 'V363', 'V364', 'V367', 'V372', 'V372A', 'V375A', 'V376', 'V376A', 'V379', 'V380']]
rec2_1 

CASEID  V201  V218                     V301  \
0            000100201  2   2.0   2.0  No conoce ningún método   
1            000100201  3   0.0   0.0                      NaN   
2            000102801  2   3.0   3.0  No conoce ningún método   
3            000102801  6   3.0   3.0  No conoce ningún método   
4            000104801  2   3.0   3.0  No conoce ningún método   
...                   ...   ...   ...                      ...   
36917        325406201  2   5.0   5.0  No conoce ningún método   
36918        325406301  2   5.0   5.0  No conoce ningún método   
36919        325407001  2   8.0   7.0  No conoce ningún método   
36920        325407201  2   2.0   1.0  No conoce ningún método   
36921        325407401  2   1.0   1.0  No conoce ningún método   

                        V302 V323 V323A V325A                   V326  \
0           Metodos modernos  NaN   NaN   NaN        Farmacia/botica   
1                        NaN  NaN   NaN   NaN                    NaN   
2           Metodos modernos  NaN   NaN   NaN       Hospital ESSALUD   
3      Metodos tradicionales  NaN   NaN   NaN                    NaN   
4           Metodos modernos  NaN   NaN   NaN  Puesto de salud MINSA   
...                      ...  ...   ...   ...                    ...   
36917       Metodos modernos  NaN   NaN   NaN  Puesto de salud MINSA   
36918       Metodos modernos  NaN   NaN   NaN                    NaN   
36919       Metodos modernos  NaN   NaN   NaN                    NaN   
36920       Metodos modernos  NaN   NaN   NaN                    NaN   
36921       Metodos modernos  NaN   NaN   NaN  Puesto de salud MINSA   

                                                    V327  ... V363  \
0                                        Farmacia/botica  ...  NaN   
1                                                    NaN  ...  NaN   
2      Hospital, Centro de Salud, Puesto de Salud, Po...  ...  NaN   
3                                                    NaN  ...  NaN   
4      Hospital, Centro de Salud, Puesto de Salud, Po...  ...  NaN   
...                                                  ...  ...  ...   
36917  Hospital, Centro de Salud, Puesto de Salud, Po...  ...  NaN   
36918                                                NaN  ...  NaN   
36919                                                NaN  ...  NaN   
36920                                                NaN  ...  NaN   
36921  Hospital, Centro de Salud, Puesto de Salud, Po...  ...  NaN   

                               V364         V367 V372 V372A       V375A  \
0          Usando el método moderno     Entonces  NaN   NaN         NaN   
1                               NaN          NaN  NaN   NaN         NaN   
2          Usando el método moderno  Esperar más  NaN   NaN         NaN   
3          No tiene la intención de          NaN  NaN   NaN  Menopausia   
4          Usando el método moderno  Esperar más  NaN   NaN         NaN   
...                             ...          ...  ...   ...         ...   
36917      Usando el método moderno  Esperar más  NaN   NaN         NaN   
36918  Usando el método tradicional     Entonces  NaN   NaN         NaN   
36919  Usando el método tradicional          NaN  NaN   NaN         NaN   
36920  Usando el método tradicional          NaN  NaN   NaN         NaN   
36921      Usando el método moderno  Esperar más  NaN   NaN         NaN   

                     V376 V376A V379 V380  
0                     NaN   NaN  NaN  NaN  
1                     NaN   NaN  NaN  NaN  
2                     NaN   NaN  NaN  NaN  
3      Infertilidad mujer   NaN  NaN  NaN  
4                     NaN   NaN  NaN  NaN  
...                   ...   ...  ...  ...  
36917                 NaN   NaN  NaN  NaN  
36918                 NaN   NaN  NaN  NaN  
36919                 NaN   NaN  NaN  NaN  
36920                 NaN   NaN  NaN  NaN  
36921                 NaN   NaN  NaN  NaN  

[36922 rows x 25 columns]

In [43]:
rec3_1 = rec_3.loc[:, ['CASEID', 'V501', 'V502', 'V503', 'V504', 'V505', 'V506', 'V507', 'V508', 'V509', 'V510', 'V511', 'V512', 'V513', 'V525', 'V613', 'V714', 'V715']]
rec3_1

CASEID             V501                  V502  \
0            000100201  2           Casado    Actualmente casada   
1            000102801  2  Viviendo juntos    Actualmente casada   
2            000102801  6  No viven juntos  Anteriormente casada   
3            000104801  2  Viviendo juntos    Actualmente casada   
4            000113601  2  Viviendo juntos    Actualmente casada   
...                   ...              ...                   ...   
33306        325406201  2  Viviendo juntos    Actualmente casada   
33307        325406301  2  Viviendo juntos    Actualmente casada   
33308        325407001  2  Viviendo juntos    Actualmente casada   
33309        325407201  2  Viviendo juntos    Actualmente casada   
33310        325407401  2  Viviendo juntos    Actualmente casada   

                 V503           V504 V505 V506  V507    V508    V509  \
0             Una vez  Vive con ella  NaN  NaN   1.0  2008.0  1297.0   
1             Una vez  Vive con ella  NaN  NaN  12.0  2011.0  1344.0   
2             Una vez            NaN  NaN  NaN   8.0  1984.0  1016.0   
3             Una vez  Vive con ella  NaN  NaN  12.0  2009.0  1320.0   
4      Mas de una vez  Vive con ella  NaN  NaN  12.0  2005.0  1272.0   
...               ...            ...  ...  ...   ...     ...     ...   
33306  Mas de una vez  Vive con ella  NaN  NaN  12.0  1985.0  1032.0   
33307         Una vez  Vive con ella  NaN  NaN  12.0  2002.0  1236.0   
33308         Una vez  Vive con ella  NaN  NaN  11.0  1986.0  1043.0   
33309         Una vez  Vive con ella  NaN  NaN   5.0  2008.0  1301.0   
33310         Una vez  Vive con ella  NaN  NaN   6.0  2013.0  1362.0   

                         V510  V511  V512              V513  V525 V613 V714  \
0             Tiene mes y año  21.0  11.0   De 10 a 14 años  17.0  2.0   Sí   
1             Tiene mes y año  28.0   7.0     De 5 a 9 años  18.0  3.0   Sí   
2      Tiene año - imputa mes  14.0  34.0  De 30 a más años  14.0  0.0   No   
3             Tiene mes y año  18.0   9.0     De 5 a 9 años  15.0  2.0   No   
4             Tiene mes y año  17.0  13.0   De 10 a 14 años  15.0  2.0   Sí   
...                       ...   ...   ...               ...   ...  ...  ...   
33306         Tiene mes y año  14.0  34.0  De 30 a más años  12.0  3.0   No   
33307         Tiene mes y año  14.0  17.0   De 15 a 19 años  12.0  3.0   No   
33308         Tiene mes y año  13.0  33.0  De 30 a más años  12.0  4.0   No   
33309         Tiene mes y año  13.0  11.0   De 10 a 14 años  13.0  2.0   No   
33310         Tiene mes y año  16.0   6.0     De 5 a 9 años  15.0  2.0   No   

       V715  
0      11.0  
1      14.0  
2       6.0  
3       6.0  
4      16.0  
...     ...  
33306   5.0  
33307   7.0  
33308   5.0  
33309  11.0  
33310  11.0  

[33311 rows x 18 columns]

In [44]:
rec1_1.attrs[ 'new_value_labels1' ] = labels_rec1
rec1_1.attrs[ 'new_var_labels1' ] = metadata1.varLabels
rec1_1.attrs[ 'new_value_labels1' ]

{'V013': {0.0: 'De 12 a 14 años de edad',
  1.0: 'De 15 a 19 años de edad',
  2.0: 'De 20 a 24 años de edad',
  3.0: 'De 25 a 29 años de edad',
  4.0: 'De 30 a 34 años de edad',
  5.0: 'De 35 a 39 años de edad',
  6.0: 'De 40 a 44 años de edad',
  7.0: 'De 45 a 49 años de edad'},
 'V014': {1.0: 'Mes y año',
  2.0: 'Mes y edad - año imputado',
  3.0: 'Año y edad - mes imputado',
  4.0: 'Año y edad - año ignorado supera el año de entrevista',
  5.0: 'Año - edad y mes imputados',
  6.0: 'Edad - año y mes imputados',
  7.0: 'Mes - edad y año imputados',
  8.0: 'Ninguno - edad, mes y año imputados'},
 'V015': {1.0: 'Completa',
  2.0: 'Ausente',
  3.0: 'Aplazada',
  4.0: 'Rechazada',
  5.0: 'Incoompleta',
  6.0: 'Discapacitada',
  7.0: 'Otra'},
 'V018': {0.0: 'Sin calendario'},
 'V019': {0.0: 'Sin calendario'},
 'V020': {0.0: 'Toda la mujer muestra', 1.0: 'Alguna muestra casada'},
 'V023': {1.0: 'Amazonas',
  2.0: 'Ancash',
  3.0: 'Apurimac',
  4.0: 'Arequipa',
  5.0: 'Ayacucho',
  6.0: 'Caj

In [45]:
rec2_1.attrs[ 'new_value_labels2' ] = labels_rec2
rec2_1.attrs[ 'new_var_labels2' ] = metadata2.varLabels
rec2_1.attrs[ 'new_value_labels2' ]

{'V213': {0.0: 'No o no está segura', 1.0: 'Sí'},
 'V215': {994.0: 'En la menopausia',
  995.0: 'Antes del último nacimiento',
  996.0: 'Nunca menstruado',
  997.0: 'Inconsistente',
  998.0: 'No sabe'},
 'V216': {0.0: 'No', 1.0: 'Si'},
 'V217': {1.0: 'Durante su periodo',
  2.0: 'Después de su período',
  3.0: 'En medio del ciclo',
  4.0: 'Antes del período',
  5.0: 'En cualquier momento',
  6.0: 'Otro',
  8.0: 'No sabe'},
 'V220': {6.0: '6+'},
 'V221': {996.0: 'Intervalo Negativo'},
 'V223': {7.0: 'Mes - fecha exacta', 8.0: 'Ninguno - fecha de colocación'},
 'V225': {1.0: 'En ese momento',
  2.0: 'Querìa esperar',
  3.0: 'No querìa tener (màs) hijas/hijos'},
 'V226': {994.0: 'En la menopausia',
  995.0: 'Antes del último nacimiento',
  996.0: 'Nunca menstruado',
  997.0: 'Inconsistente',
  998.0: 'No sabe'},
 'V227': {0.0: 'No marcado',
  1.0: 'N> int. desde el nacimiento',
  2.0: 'N + amen > intervalo',
  3.0: 'N, pero per no ret',
  4.0: '95, pero sin nacimiento',
  5.0: '95, pero p

In [46]:
rec3_1.attrs[ 'new_value_labels3' ] = labels_rec3
rec3_1.attrs[ 'new_var_labels3' ] = metadata3.varLabels
rec3_1.attrs[ 'new_value_labels3' ]

{'V501': {0.0: 'Nunca casada',
  1.0: 'Casado',
  2.0: 'Viviendo juntos',
  3.0: 'Viuda',
  4.0: 'Divorciada',
  5.0: 'No viven juntos'},
 'V502': {0.0: 'Nunca casada',
  1.0: 'Actualmente casada',
  2.0: 'Anteriormente casada'},
 'V503': {1.0: 'Una vez', 2.0: 'Mas de una vez'},
 'V504': {1.0: 'Vive con ella', 2.0: 'Vive en otro sitio'},
 'V505': {0.0: 'No otras esposas', 98.0: 'No sabe'},
 'V506': {98.0: 'No sabe'},
 'V510': {1.0: 'Tiene mes y año',
  2.0: 'Tiene mes y edad - imputa año',
  3.0: 'Tiene año y edad - imputa mes',
  5.0: 'Tiene año - imputa mes',
  6.0: 'Tiene edad - imputa mes año',
  7.0: 'Tiene mes - imputa año',
  8.0: 'Ninguno - imputa todo'},
 'V513': {0.0: 'Nunca casada',
  1.0: 'De 0 a 4 años',
  2.0: 'De 5 a 9 años',
  3.0: 'De 10 a 14 años',
  4.0: 'De 15 a 19 años',
  5.0: 'De 20 a 24 años',
  6.0: 'De 25 a 29 años',
  7.0: 'De 30 a más años'},
 'V525': {0.0: 'Nunca tuvo relaciones sexuales', 96.0: 'En la primera union'},
 'V527': {100.0: 'Menos de un día',
  

3. Generate a new column for `rec1_1` named as `year`. It should be equal to `2019`. Also, you must update this new variable for the `var_labels` dictionary. Generate a new key for `new_var_labels1` and the value for this key should be **"Year of the survey"** **Hint: Use `loc` and `update` method.**

## Solution


In [47]:
rec1_1.loc[:, 'year'] = 2019

metadata1.varLabels.update({'year': "Year of the survey"})

# Update the attribute for var_labels1 in the DataFrame
rec1_1.attrs['new_var_labels1'] = metadata1.varLabels
rec1_1.attrs['new_var_labels1']

{'ID1': 'Año',
 'HHID': 'Identificación Cuestionario del Hogar',
 'CASEID': 'Identificación Cuestionario Individual',
 'V001': 'Conglomerado',
 'V002': 'Número de vivienda',
 'V003': 'Número de línea de entrevistada',
 'V004': 'Unidad de área final',
 'V007': 'Año de la entrevista',
 'V008': 'Fecha de la entrevista, Codificación centenaria de meses (CMC)',
 'V009': 'Mes de nacimiento de la entrevistada',
 'V010': 'Año de nacimiento de la entrevistada',
 'V011': 'Fecha de nacimiento, Codificación centenaria de meses (CMC)',
 'V012': 'Edad actual - entrevistada',
 'V013': 'Edad actual por grupos de 5 años',
 'V014': 'Integridad de la información para la fecha de nacimiento ',
 'V015': 'Resultado entrevista individual',
 'V017': 'Inicio del calendario, Codificación centenaria de mesesl CMC',
 'V018': 'Columna del mes de la entrevista',
 'V019': 'Duración del calendario',
 'V019A': 'Número de columnas de calendario',
 'V020': 'Muestra alguna vez casada',
 'V021': 'Unidad de muestreo primar

4. Merge `rec1_1`, `rec2_1`, and `rec3_1` using **CASEID**. Name this new object as `endes_2019`. **Hint: Use [this link](https://stackoverflow.com/questions/53645882/pandas-merging-101)**

In [48]:
# First, use merge "outer" to conserve Non matched keys between rec1_1 and rec_2_1
endes_2019 = rec1_1.merge(rec2_1, on='CASEID', how='outer')


In [49]:
# Then, we merge the remaining dataset rec3_1
endes_2019 = endes_2019.merge(rec3_1, on='CASEID', how='outer')

In [50]:
endes_2019

CASEID V000    V001  V002  V003    V004    V007    V008  \
0            000100201  2  PE6     1.0   2.0   2.0     1.0  2019.0  1434.0   
1            000100201  3  PE6     1.0   2.0   3.0     1.0  2019.0  1434.0   
2            000102801  2  PE6     1.0  28.0   2.0     1.0  2019.0  1434.0   
3            000102801  6  PE6     1.0  28.0   6.0     1.0  2019.0  1434.0   
4            000104801  2  PE6     1.0  48.0   2.0     1.0  2019.0  1434.0   
...                   ...  ...     ...   ...   ...     ...     ...     ...   
38330        325406201  2  PE6  3254.0  62.0   2.0  3254.0  2019.0  1440.0   
38331        325406301  2  PE6  3254.0  63.0   2.0  3254.0  2019.0  1440.0   
38332        325407001  2  PE6  3254.0  70.0   2.0  3254.0  2019.0  1440.0   
38333        325407201  2  PE6  3254.0  72.0   2.0  3254.0  2019.0  1440.0   
38334        325407401  2  PE6  3254.0  74.0   2.0  3254.0  2019.0  1440.0   

       V009    V010  ...    V508    V509                    V510  V511  V512  \
0       4.0  1986.0  ...  2008.0  1297.0         Tiene mes y año  21.0  11.0   
1       1.0  2007.0  ...     NaN     NaN                     NaN   NaN   NaN   
2       6.0  1983.0  ...  2011.0  1344.0         Tiene mes y año  28.0   7.0   
3       3.0  1970.0  ...  1984.0  1016.0  Tiene año - imputa mes  14.0  34.0   
4       5.0  1991.0  ...  2009.0  1320.0         Tiene mes y año  18.0   9.0   
...     ...     ...  ...     ...     ...                     ...   ...   ...   
38330  12.0  1971.0  ...  1985.0  1032.0         Tiene mes y año  14.0  34.0   
38331   6.0  1988.0  ...  2002.0  1236.0         Tiene mes y año  14.0  17.0   
38332   7.0  1973.0  ...  1986.0  1043.0         Tiene mes y año  13.0  33.0   
38333  12.0  1994.0  ...  2008.0  1301.0         Tiene mes y año  13.0  11.0   
38334  10.0  1996.0  ...  2013.0  1362.0         Tiene mes y año  16.0   6.0   

                   V513  V525 V613 V714  V715  
0       De 10 a 14 años  17.0  2.0   Sí  11.0  
1                   NaN   NaN  NaN  NaN   NaN  
2         De 5 a 9 años  18.0  3.0   Sí  14.0  
3      De 30 a más años  14.0  0.0   No   6.0  
4         De 5 a 9 años  15.0  2.0   No   6.0  
...                 ...   ...  ...  ...   ...  
38330  De 30 a más años  12.0  3.0   No   5.0  
38331   De 15 a 19 años  12.0  3.0   No   7.0  
38332  De 30 a más años  12.0  4.0   No   5.0  
38333   De 10 a 14 años  13.0  2.0   No  11.0  
38334     De 5 a 9 años  15.0  2.0   No  11.0  

[38335 rows x 64 columns]

The number of rows is 38335, exactly the same number of rows as in rec1_1, the largest of the three datasets. \
The number of columns is 64, 22+25+18-2+1 = 64. The merge is correct.

5. Unify all the `new_var_labels` in one object and `new_value_labels` in another one object. Name these two objects as `var_labels` and `value_labels`. Use them to generate new attributes for `endes_2019`. These attributes should be named as `var_labels` and `value_labels`. **Hint: Use `update` method.**

In [51]:
# Create an empty dictionary to combine the three new_var_labels
var_labels = {}

# Use update method to merge dictionaries
var_labels.update(rec1_1.attrs['new_var_labels1'])
var_labels.update(rec2_1.attrs['new_var_labels2'])
var_labels.update(rec3_1.attrs['new_var_labels3'])

#verify the new dict
var_labels

{'ID1': 'Año',
 'HHID': 'Identificación Cuestionario del Hogar',
 'CASEID': 'Identificación Cuestionario Individual',
 'V001': 'Conglomerado',
 'V002': 'Número de vivienda',
 'V003': 'Número de línea de entrevistada',
 'V004': 'Unidad de área final',
 'V007': 'Año de la entrevista',
 'V008': 'Fecha de la entrevista, Codificación centenaria de meses (CMC)',
 'V009': 'Mes de nacimiento de la entrevistada',
 'V010': 'Año de nacimiento de la entrevistada',
 'V011': 'Fecha de nacimiento, Codificación centenaria de meses (CMC)',
 'V012': 'Edad actual - entrevistada',
 'V013': 'Edad actual por grupos de 5 años',
 'V014': 'Integridad de la información para la fecha de nacimiento ',
 'V015': 'Resultado entrevista individual',
 'V017': 'Inicio del calendario, Codificación centenaria de mesesl CMC',
 'V018': 'Columna del mes de la entrevista',
 'V019': 'Duración del calendario',
 'V019A': 'Número de columnas de calendario',
 'V020': 'Muestra alguna vez casada',
 'V021': 'Unidad de muestreo primar

In [52]:
# Create an empty dictionary to combine the three new_value_labels
value_labels = {}

# Use update method to merge dictionaries
value_labels.update(rec1_1.attrs['new_value_labels1'])
value_labels.update(rec2_1.attrs['new_value_labels2'])
value_labels.update(rec3_1.attrs['new_value_labels3'])

#verify the new dict
value_labels

{'V013': {0.0: 'De 12 a 14 años de edad',
  1.0: 'De 15 a 19 años de edad',
  2.0: 'De 20 a 24 años de edad',
  3.0: 'De 25 a 29 años de edad',
  4.0: 'De 30 a 34 años de edad',
  5.0: 'De 35 a 39 años de edad',
  6.0: 'De 40 a 44 años de edad',
  7.0: 'De 45 a 49 años de edad'},
 'V014': {1.0: 'Mes y año',
  2.0: 'Mes y edad - año imputado',
  3.0: 'Año y edad - mes imputado',
  4.0: 'Año y edad - año ignorado supera el año de entrevista',
  5.0: 'Año - edad y mes imputados',
  6.0: 'Edad - año y mes imputados',
  7.0: 'Mes - edad y año imputados',
  8.0: 'Ninguno - edad, mes y año imputados'},
 'V015': {1.0: 'Completa',
  2.0: 'Ausente',
  3.0: 'Aplazada',
  4.0: 'Rechazada',
  5.0: 'Incoompleta',
  6.0: 'Discapacitada',
  7.0: 'Otra'},
 'V018': {0.0: 'Sin calendario'},
 'V019': {0.0: 'Sin calendario'},
 'V020': {0.0: 'Toda la mujer muestra', 1.0: 'Alguna muestra casada'},
 'V023': {1.0: 'Amazonas',
  2.0: 'Ancash',
  3.0: 'Apurimac',
  4.0: 'Arequipa',
  5.0: 'Ayacucho',
  6.0: 'Caj

In [53]:
#update var labels and value labels for endes_2019 dataset
endes_2019.attrs.update({'var_labels': var_labels, 'value_labels': value_labels})

In [54]:
#verify
print(endes_2019.attrs['var_labels'])
print(endes_2019.attrs['value_labels'])


{'ID1': 'Año', 'HHID': 'Identificación Cuestionario del Hogar', 'CASEID': 'Identificación Cuestionario Individual', 'V001': 'Conglomerado', 'V002': 'Número de vivienda', 'V003': 'Número de línea de entrevistada', 'V004': 'Unidad de área final', 'V007': 'Año de la entrevista', 'V008': 'Fecha de la entrevista, Codificación centenaria de meses (CMC)', 'V009': 'Mes de nacimiento de la entrevistada', 'V010': 'Año de nacimiento de la entrevistada', 'V011': 'Fecha de nacimiento, Codificación centenaria de meses (CMC)', 'V012': 'Edad actual - entrevistada', 'V013': 'Edad actual por grupos de 5 años', 'V014': 'Integridad de la información para la fecha de nacimiento ', 'V015': 'Resultado entrevista individual', 'V017': 'Inicio del calendario, Codificación centenaria de mesesl CMC', 'V018': 'Columna del mes de la entrevista', 'V019': 'Duración del calendario', 'V019A': 'Número de columnas de calendario', 'V020': 'Muestra alguna vez casada', 'V021': 'Unidad de muestreo primario - conglomerado', '

6. Get the min, max, sd, n_obs, n_missing for the following columns **total children ever born (V201)**, **Ideal number of children (V613)**, **Husbands education-single yrs (V715)**, and **Age at first marriage (V511)**. We want a dataframe with the following columns **Variables, Min, Max, Mean, N_obs, N_missing** and sort by the number of missing rows. **Hint: Use `describe` and `pivot` methods**

In [55]:
#select the relevant columns
columns = ['V201', 'V613', 'V715', 'V511']

#extract the columns as a new dataset
statistics_dataset = endes_2019.loc[:, ['V201', 'V613', 'V715', 'V511']]

#verify Dtype of the columns
statistics_dataset.info()

#Since we have 2 catecorical variables we have to transform them to numeric to run statistics

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38335 entries, 0 to 38334
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   V201    36922 non-null  float64 
 1   V613    33311 non-null  category
 2   V715    25881 non-null  category
 3   V511    25881 non-null  float64 
dtypes: category(2), float64(2)
memory usage: 974.7 KB


In [56]:
#Transform the two columns into numeric type
statistics_dataset['V613'] = pd.to_numeric(statistics_dataset['V613'], errors='coerce')
statistics_dataset['V715'] = pd.to_numeric(statistics_dataset['V715'], errors='coerce')


#run summary
#summary_stats = statistics_dataset.describe().transpose()
summary_stats = statistics_dataset.describe()
summary_stats

V201          V613          V715          V511
count  36922.000000  33260.000000  25626.000000  25881.000000
mean       1.781946      2.311906     10.764965     20.463738
std        1.749109      1.183034      3.578349      4.858358
min        0.000000      0.000000      0.000000     10.000000
25%        0.000000      2.000000      9.000000     17.000000
50%        1.000000      2.000000     11.000000     19.000000
75%        3.000000      3.000000     14.000000     23.000000
max       13.000000     20.000000     17.000000     48.000000

In [57]:
# Use "pivot_table" to put index as column names and column names as values
summary_stats = summary_stats.pivot_table(index=None, columns=summary_stats.index)
summary_stats


25%   50%   75%    count   max       mean   min       std
V201   0.0   1.0   3.0  36922.0  13.0   1.781946   0.0  1.749109
V511  17.0  19.0  23.0  25881.0  48.0  20.463738  10.0  4.858358
V613   2.0   2.0   3.0  33260.0  20.0   2.311906   0.0  1.183034
V715   9.0  11.0  14.0  25626.0  17.0  10.764965   0.0  3.578349

In [58]:
# Order the dataframe 
result_df = pd.DataFrame({
    'Variables': summary_stats.index,
    'Min': summary_stats['min'],
    'Max': summary_stats['max'],
    'Mean': summary_stats['mean'],
    'N_obs': summary_stats['count'],
    'N_missing': statistics_dataset[columns].isnull().sum()
})
result_df

Variables   Min   Max       Mean    N_obs  N_missing
V201      V201   0.0  13.0   1.781946  36922.0       1413
V511      V511  10.0  48.0  20.463738  25881.0      12454
V613      V613   0.0  20.0   2.311906  33260.0       5075
V715      V715   0.0  17.0  10.764965  25626.0      12709

7. Use `endes_2019` data to generate a new object named `mean_key_vars` to find the mean of **total children ever born (V201)**, **Ideal number of children (V613)**, **Husbands education-single yrs (V715)**, and **Age at first marriage (V511)** by year and department **(V024)**. Name these columns as **mean_total_children, mean_ideal_children, mean_hb_yr_educ and mean_first_marriage**, respectively. **Hint: Use groupby and [this link](https://stackoverflow.com/questions/40901770/is-there-a-simple-way-to-change-a-column-of-yes-no-to-1-0-in-a-pandas-dataframe).**

In [59]:
# Select the variables
endes_2019_new = endes_2019[['V024', 'V201', 'V613', 'V715', 'V511']]

# Rename the columns
endes_2019_new.columns = ['department', 'total_children', 'ideal_children', 'hb_yr_educ', 'first_marriage']
endes_2019_new

department  total_children ideal_children hb_yr_educ  first_marriage
0       Amazonas             2.0            2.0       11.0            21.0
1       Amazonas             0.0            NaN        NaN             NaN
2       Amazonas             3.0            3.0       14.0            28.0
3       Amazonas             3.0            0.0        6.0            14.0
4       Amazonas             3.0            2.0        6.0            18.0
...          ...             ...            ...        ...             ...
38330    Ucayali             5.0            3.0        5.0            14.0
38331    Ucayali             5.0            3.0        7.0            14.0
38332    Ucayali             8.0            4.0        5.0            13.0
38333    Ucayali             2.0            2.0       11.0            13.0
38334    Ucayali             1.0            2.0       11.0            16.0

[38335 rows x 5 columns]

We use the groupby function to summarize the values (obtain means).

In [60]:
# Convert variables to numerical
cols = ['total_children', 'ideal_children', 'hb_yr_educ', 'first_marriage']
endes_2019_new[cols] = endes_2019_new[cols].apply(pd.to_numeric, errors='coerce')

# Calculate averages by year and department
mean_key_vars = endes_2019_new.groupby(['department']).mean()
mean_key_vars
# Rename columns
mean_key_vars.columns = ['mean_total_children', 'mean_ideal_children', 'mean_hb_yr_educ', 'mean_first_marriage']
mean_key_vars

C:\Users\DDelgado\AppData\Local\Temp\ipykernel_29772\1741609205.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  endes_2019_new[cols] = endes_2019_new[cols].apply(pd.to_numeric, errors='coerce')


mean_total_children  mean_ideal_children  mean_hb_yr_educ  \
department                                                                  
 Huanuco                   1.989754             2.446940         8.842528   
 Ica                       1.671034             2.353429        11.825332   
 Junin                     1.836656             2.199660        10.942122   
 La Libertad               1.704082             2.355519         9.907909   
 Lambayeque                1.673928             2.399849        11.123396   
 Lima                      1.476121             2.235107        11.731348   
 Loreto                    2.157568             2.476703        10.109375   
 Madre de Dios             1.972514             2.272398        10.729053   
 Moquegua                  1.626724             1.965613        12.568605   
 Pasco                     1.854436             2.526570        11.032680   
 Piura                     1.812752             2.611650        10.032480   
 Puno                      1.794307             2.024316        10.836151   
 San Martin                1.903340             2.493333         9.621053   
 Tacna                     1.436502             2.013083        12.141403   
 Tumbes                    1.756925             2.469824        11.118095   
 Ucayali                   2.012258             2.388399        10.709355   
Amazonas                   1.980892             2.443906         9.440580   
Ancash                     1.771065             2.383260        10.495381   
Apurimac                   2.003463             2.207269        10.537530   
Arequipa                   1.539910             2.137875        12.244275   
Ayacucho                   1.848787             2.209026        10.470772   
Cajamarca                  1.931559             2.545923         8.708874   
Callao                     1.567164             2.241724        11.663909   
Cusco                      1.894415             2.100097        10.569194   
Huancavelica               2.083969             2.295098         9.879221   

                mean_first_marriage  
department                           
 Huanuco                  19.818632  
 Ica                      20.747208  
 Junin                    20.683761  
 La Libertad              20.774678  
 Lambayeque               20.953831  
 Lima                     21.946329  
 Loreto                   18.448529  
 Madre de Dios            19.467449  
 Moquegua                 21.640046  
 Pasco                    20.395379  
 Piura                    20.451737  
 Puno                     20.398964  
 San Martin               18.953155  
 Tacna                    22.065095  
 Tumbes                   19.409821  
 Ucayali                  18.985481  
Amazonas                  19.266152  
Ancash                    20.591581  
Apurimac                  20.064982  
Arequipa                  22.304394  
Ayacucho                  20.340228  
Cajamarca                 19.610695  
Callao                    21.517241  
Cusco                     20.283154  
Huancavelica              19.780769

8. Reshape `mean_key_vars` from wide to long. We want a dataframe with three columns **dpto, variables, values**. Name this object as `reshape_mean_key_vars`. **Hint: Use melt method**.

In [61]:
mean_key_vars['department'] = mean_key_vars.index
mean_key_vars

mean_total_children  mean_ideal_children  mean_hb_yr_educ  \
department                                                                  
 Huanuco                   1.989754             2.446940         8.842528   
 Ica                       1.671034             2.353429        11.825332   
 Junin                     1.836656             2.199660        10.942122   
 La Libertad               1.704082             2.355519         9.907909   
 Lambayeque                1.673928             2.399849        11.123396   
 Lima                      1.476121             2.235107        11.731348   
 Loreto                    2.157568             2.476703        10.109375   
 Madre de Dios             1.972514             2.272398        10.729053   
 Moquegua                  1.626724             1.965613        12.568605   
 Pasco                     1.854436             2.526570        11.032680   
 Piura                     1.812752             2.611650        10.032480   
 Puno                      1.794307             2.024316        10.836151   
 San Martin                1.903340             2.493333         9.621053   
 Tacna                     1.436502             2.013083        12.141403   
 Tumbes                    1.756925             2.469824        11.118095   
 Ucayali                   2.012258             2.388399        10.709355   
Amazonas                   1.980892             2.443906         9.440580   
Ancash                     1.771065             2.383260        10.495381   
Apurimac                   2.003463             2.207269        10.537530   
Arequipa                   1.539910             2.137875        12.244275   
Ayacucho                   1.848787             2.209026        10.470772   
Cajamarca                  1.931559             2.545923         8.708874   
Callao                     1.567164             2.241724        11.663909   
Cusco                      1.894415             2.100097        10.569194   
Huancavelica               2.083969             2.295098         9.879221   

                mean_first_marriage      department  
department                                           
 Huanuco                  19.818632         Huanuco  
 Ica                      20.747208             Ica  
 Junin                    20.683761           Junin  
 La Libertad              20.774678     La Libertad  
 Lambayeque               20.953831      Lambayeque  
 Lima                     21.946329            Lima  
 Loreto                   18.448529          Loreto  
 Madre de Dios            19.467449   Madre de Dios  
 Moquegua                 21.640046        Moquegua  
 Pasco                    20.395379           Pasco  
 Piura                    20.451737           Piura  
 Puno                     20.398964            Puno  
 San Martin               18.953155      San Martin  
 Tacna                    22.065095           Tacna  
 Tumbes                   19.409821          Tumbes  
 Ucayali                  18.985481         Ucayali  
Amazonas                  19.266152        Amazonas  
Ancash                    20.591581          Ancash  
Apurimac                  20.064982        Apurimac  
Arequipa                  22.304394        Arequipa  
Ayacucho                  20.340228        Ayacucho  
Cajamarca                 19.610695       Cajamarca  
Callao                    21.517241          Callao  
Cusco                     20.283154           Cusco  
Huancavelica              19.780769    Huancavelica

In [62]:
reshape_mean_key_vars = pd.melt(mean_key_vars, id_vars=['department'], value_vars=['mean_total_children', 'mean_ideal_children', 'mean_hb_yr_educ', 'mean_first_marriage'])
reshape_mean_key_vars

department             variable      value
0        Huanuco  mean_total_children   1.989754
1            Ica  mean_total_children   1.671034
2          Junin  mean_total_children   1.836656
3    La Libertad  mean_total_children   1.704082
4     Lambayeque  mean_total_children   1.673928
..           ...                  ...        ...
95      Ayacucho  mean_first_marriage  20.340228
96     Cajamarca  mean_first_marriage  19.610695
97        Callao  mean_first_marriage  21.517241
98         Cusco  mean_first_marriage  20.283154
99  Huancavelica  mean_first_marriage  19.780769

[100 rows x 3 columns]

9. Replicate your answers for questions 7 and 8, but in one line of code. Make it the most simple as possible. **NO HINT**

We used ChatGPT to replicate our previous steps in one line of code.

In [63]:
result = (endes_2019[['V024', 'V201', 'V613', 'V715', 'V511']]
    .rename(columns={'V024': 'department', 'V201': 'total_children', 'V613': 'ideal_children', 'V715': 'hb_yr_educ', 'V511': 'first_marriage'})
    .assign(total_children=pd.to_numeric(endes_2019['V201'], errors='coerce'),
            ideal_children=pd.to_numeric(endes_2019['V613'], errors='coerce'),
            hb_yr_educ=pd.to_numeric(endes_2019['V715'], errors='coerce'),
            first_marriage=pd.to_numeric(endes_2019['V511'], errors='coerce'))
    .groupby(['department'])
    .mean()
    .rename(columns={'total_children': 'mean_total_children', 'ideal_children': 'mean_ideal_children', 'hb_yr_educ': 'mean_hb_yr_educ', 'first_marriage': 'mean_first_marriage'})
    .assign(department=lambda x: x.index)
    .melt(id_vars=['department'], value_vars=['mean_total_children', 'mean_ideal_children', 'mean_hb_yr_educ', 'mean_first_marriage'])
)

result

department             variable      value
0        Huanuco  mean_total_children   1.989754
1            Ica  mean_total_children   1.671034
2          Junin  mean_total_children   1.836656
3    La Libertad  mean_total_children   1.704082
4     Lambayeque  mean_total_children   1.673928
..           ...                  ...        ...
95      Ayacucho  mean_first_marriage  20.340228
96     Cajamarca  mean_first_marriage  19.610695
97        Callao  mean_first_marriage  21.517241
98         Cusco  mean_first_marriage  20.283154
99  Huancavelica  mean_first_marriage  19.780769

[100 rows x 3 columns]

10. Merge `reshape_mean_key_vars` with `endes_2019`. Name this object `final_result`. **Hint: Use merge.**

Since we have a non-unique ID datasets we have to perform a many-to-many merge. We accomplish this by using the pd.merge function.

In [64]:
reshape_mean_key_vars.rename(columns={'department': 'V024'}, inplace=True)
merged_data = pd.merge(endes_2019,reshape_mean_key_vars , on='V024', how='outer')
merged_data

CASEID V000    V001  V002  V003    V004    V007    V008  \
0             000100201  2  PE6     1.0   2.0   2.0     1.0  2019.0  1434.0   
1             000100201  2  PE6     1.0   2.0   2.0     1.0  2019.0  1434.0   
2             000100201  2  PE6     1.0   2.0   2.0     1.0  2019.0  1434.0   
3             000100201  2  PE6     1.0   2.0   2.0     1.0  2019.0  1434.0   
4             000100201  3  PE6     1.0   2.0   3.0     1.0  2019.0  1434.0   
...                    ...  ...     ...   ...   ...     ...     ...     ...   
153335        325407201  2  PE6  3254.0  72.0   2.0  3254.0  2019.0  1440.0   
153336        325407401  2  PE6  3254.0  74.0   2.0  3254.0  2019.0  1440.0   
153337        325407401  2  PE6  3254.0  74.0   2.0  3254.0  2019.0  1440.0   
153338        325407401  2  PE6  3254.0  74.0   2.0  3254.0  2019.0  1440.0   
153339        325407401  2  PE6  3254.0  74.0   2.0  3254.0  2019.0  1440.0   

        V009    V010  ...             V510  V511  V512             V513  V525  \
0        4.0  1986.0  ...  Tiene mes y año  21.0  11.0  De 10 a 14 años  17.0   
1        4.0  1986.0  ...  Tiene mes y año  21.0  11.0  De 10 a 14 años  17.0   
2        4.0  1986.0  ...  Tiene mes y año  21.0  11.0  De 10 a 14 años  17.0   
3        4.0  1986.0  ...  Tiene mes y año  21.0  11.0  De 10 a 14 años  17.0   
4        1.0  2007.0  ...              NaN   NaN   NaN              NaN   NaN   
...      ...     ...  ...              ...   ...   ...              ...   ...   
153335  12.0  1994.0  ...  Tiene mes y año  13.0  11.0  De 10 a 14 años  13.0   
153336  10.0  1996.0  ...  Tiene mes y año  16.0   6.0    De 5 a 9 años  15.0   
153337  10.0  1996.0  ...  Tiene mes y año  16.0   6.0    De 5 a 9 años  15.0   
153338  10.0  1996.0  ...  Tiene mes y año  16.0   6.0    De 5 a 9 años  15.0   
153339  10.0  1996.0  ...  Tiene mes y año  16.0   6.0    De 5 a 9 años  15.0   

       V613 V714  V715             variable      value  
0       2.0   Sí  11.0  mean_total_children   1.980892  
1       2.0   Sí  11.0  mean_ideal_children   2.443906  
2       2.0   Sí  11.0      mean_hb_yr_educ   9.440580  
3       2.0   Sí  11.0  mean_first_marriage  19.266152  
4       NaN  NaN   NaN  mean_total_children   1.980892  
...     ...  ...   ...                  ...        ...  
153335  2.0   No  11.0  mean_first_marriage  18.985481  
153336  2.0   No  11.0  mean_total_children   2.012258  
153337  2.0   No  11.0  mean_ideal_children   2.388399  
153338  2.0   No  11.0      mean_hb_yr_educ  10.709355  
153339  2.0   No  11.0  mean_first_marriage  18.985481  

[153340 rows x 66 columns]